### PR TITLE
Remove replace_all and make VTs mutable

### PR DESCRIPTION
**Pull request by mlazos**
1.  Removes calls to `replace_all` and `clone` and makes VTs mutable.
2. Properly handles Tuple Iterator mutation. Previously TupleIterator variables would only be properly reconstructed if they were advanced at least once in a frame. On calls to `next`, the source information would be lost (due to constructing a new iterator without using builder), which would ensure that during codegen the variable would be reconstructed from scratch. Now that VTs are mutated, the source is never lost, so we need to properly track mutation and handle it by replaying calls to `next` at the end of the modified bytecode.
3. Added test for checking iadd side effects, this was missing in our unit test coverage.
4. Fixed two incorrect sources, DelayGraphBreakVariable, and UserMethodVariable both relied on setting the source to AttrSource(parent, name) at the callsite of `var_getattr`.
5. Fixed a bug in inplace adding for lists, it would set the resulting VariableTracker's source to `None` which would utilize a different reconstruct path in codegen. Now this is handled explicitly by reconstructing vars when allow_cache=`False`, so that during side effect replay, the mutated var is correctly updated.

In subsequent PRs:
* Refactoring side effect tracking to be significantly simpler (I think we only need an `is_modified` flag)
* Refactor `next_variables` iterator to match the signature of `next`
* Remove all references to `options` in the code
* Refactor VTs representing mutable collections to implement their own mutation update handling
* Remove clone and/or make it specific to lists for creating slices
* Add mutation tracking/replay for sets
* Add mutation tracking/replay for iter.py
* Removing setting source in builder (it's set at the top level after a var is returned)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py
Comment (mlazos) [inline torch/_dynamo/variables/builtin.py:354]: Figure out what options are handled here

File: torch/_dynamo/variables/builtin.py
Comment (mlazos) [inline torch/_dynamo/variables/builtin.py:1165]: Remove comment

File: torch/_dynamo/variables/builtin.py
Comment (mlazos) [inline torch/_dynamo/variables/builtin.py:1193]: Remove comment

File: torch/_dynamo/variables/functions.py
Comment (mlazos) [inline torch/_dynamo/variables/functions.py:769]: Remove comment

File: torch/_dynamo/variables/iter.py
Comment (mlazos) [inline torch/_dynamo/variables/iter.py:26]: Remove comment

File: test/dynamo/test_misc.py
Comment (mlazos) [inline test/dynamo/test_misc.py:7578]: NB: this was unused

File: torch/_dynamo/side_effects.py
Comment (jansel) [inline torch/_dynamo/side_effects.py:515]: This assumes the user doesn't have a variable named `next`, which is unsound.

**Review by jansel — CHANGES_REQUESTED** [review]
1) How are we ensuring that there are no stray clones or manual copies left?  Can we remove the clone APIs and audit usage of apply?  Add some asserts?
2) I believe this may be blocked on #113235 -- the checkpointing in higher_order_ops won't work right with this change.

File: torch/_dynamo/side_effects.py
Comment (mlazos) [reply torch/_dynamo/side_effects.py:515]: I'll add a special dynamo util for this

**Comment by mlazos** [comment]
> 1. How are we ensuring that there are no stray clones or manual copies left?  Can we remove the clone APIs and audit usage of apply?  Add some asserts?
> 2. I believe this may be blocked on [[dynamo] Remove usage of copy_graphstate/restore_graphstate in higher_order_ops.py #113235](https://github.com/pytorch/pytorch/issues/113235) -- the checkpointing in higher_order_ops won't work right with this change.

1. I removed most usages of clone, the only remaining case is used for creating slices of a list. For apply, it's only used to realize all variables in scope for export and check liveness in side_effects. I no longer call clone on modified vars in apply, it's essentially treemap now.
2. I think we're passing the cond and hop tests in spite of this, I think we can trace through both branches since leaked side effects to outer scopes are banned. I can double check this though before merging

File: test/dynamo/test_misc.py
Comment (voznesenskym) [reply test/dynamo/test_misc.py:7578]: Separate PR

File: torch/_dynamo/symbolic_convert.py
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:667]: Running repl on side effects here, aside from doing __dict__ updating and transfer, also ensured that source was correct for when we did oldvar->newvar. Maybe this is not possible with a general removal of replace_all, but it would be worth triple checking this.

File: torch/_dynamo/utils.py
Comment (voznesenskym) [inline torch/_dynamo/utils.py:931]: rm?

Comment (mlazos) [reply torch/_dynamo/utils.py:931]: no, this is needed to ensure that next doesn't shadow a user var named next when replaying mutations. We import the fully qualified name

**Comment by voznesenskym** [comment]
Seems reasonable, nothing too weird. Would stamp if tests passed and HOP story was satisfying.

**Review by jansel — CHANGES_REQUESTED** [review]
Tests are failing and this is still blocked on #113235.

If we land this before fixing #113235 it will be pretty bad correctness issues:
1) mutate a ListVariable inside a higher order op
2) the mutation is visible outside the higher order op, and maybe reference things that don't exist in the outer scope

**Comment by mlazos** [comment]
@jansel  this is ready for review but I'll submit another PR first to remove always_restore, once that's done I'll remove the remaining calls to checkpoint here.

File: torch/_dynamo/utils.py
Comment (jansel) [inline torch/_dynamo/utils.py:947]:
```suggestion
iter_next = next

```

File: torch/_dynamo/variables/base.py
Comment (jansel) [inline torch/_dynamo/variables/base.py:190]: We should audit all remaining uses of `apply()` to ensure this isn't doing any mutations we don't want.

I worry we have something like:
```
def visit(node):
   ...

# node result is ignored
VariableTracker.apply(visit, obj)
```

Previously that would not have any side effects.  Now it does.

File: torch/_dynamo/variables/builder.py
Comment (jansel) [inline torch/_dynamo/variables/builder.py:473]: Is adding all the `source=` in this file needed (throughout the file)?  I thought source was set automatically set for VariableBuilder.

File: torch/_dynamo/variables/dicts.py
Comment (jansel) [inline torch/_dynamo/variables/dicts.py:410]: Missing call to mutation?

File: torch/_dynamo/variables/dicts.py
Comment (jansel) [inline torch/_dynamo/variables/dicts.py:401]: No call to mutation() `_add` function?

File: torch/_dynamo/variables/dicts.py
Comment (jansel) [inline torch/_dynamo/variables/dicts.py:405]: No call to mutation()?

File: torch/_dynamo/variables/functions.py
Comment (jansel) [inline torch/_dynamo/variables/functions.py:776]: I think this case mutation is actually incorrect.  You could have:
```
kernel.run(.... grid=a)
kernel.run(.... grid=b)
```
neither of these mutate `kernel`.

File: torch/_dynamo/variables/iter.py
Comment (jansel) [inline torch/_dynamo/variables/iter.py:42]: No call to mutation?

File: torch/_dynamo/variables/iter.py
Comment (jansel) [inline torch/_dynamo/variables/iter.py:74]: No call to mutation?

File: torch/_dynamo/variables/iter.py
Comment (jansel) [inline torch/_dynamo/variables/iter.py:80]: No call to mutation?

File: torch/_dynamo/variables/iter.py
Comment (jansel) [inline torch/_dynamo/variables/iter.py:83]: No call to mutation?

File: torch/_dynamo/variables/optimizer.py
Comment (jansel) [inline torch/_dynamo/variables/optimizer.py:191]: No call to mutation?

**Review by jansel — CHANGES_REQUESTED** [review]
Please take a careful pass over this PR to ensure we got all the mutation correct.

File: torch/_dynamo/variables/base.py
Comment (mlazos) [reply torch/_dynamo/variables/base.py:190]: Yeah I've taken a look, all of the uses of apply either 1) eagerly realize all vars or 2) track live vars in side effects, so we are good here.

File: torch/_dynamo/variables/dicts.py
Comment (mlazos) [reply torch/_dynamo/variables/dicts.py:410]: Hmm I'll add this, but right now I don't think sets track mutation at all since we don't replay mutations for them at the end so it's in essence a no-op. 

I can add mutation tracking for sets in a follow up PR

File: torch/_dynamo/variables/dicts.py
Comment (mlazos) [reply torch/_dynamo/variables/dicts.py:401]: So this is also called in the constructor, so I added a mutation() call in pop and add method handling.

File: torch/_dynamo/variables/dicts.py
Comment (jansel) [reply torch/_dynamo/variables/dicts.py:410]: Don't we need to assert there aren't side effects escaping the scope?

File: torch/_dynamo/variables/builder.py
Comment (mlazos) [reply torch/_dynamo/variables/builder.py:473]: Yeah you're right, I think I was considering removing that source assignment at some point when there were more tests failing. I'll remove.

Comment (mlazos) [reply torch/_dynamo/variables/builder.py:473]: Actually, it's needed now for vars where we are going to track them in side effects. I assert that all vars you pass to it have a source which makes sense. This assert wasn't there previously.

File: torch/_dynamo/variables/functions.py
Comment (mlazos) [reply torch/_dynamo/variables/functions.py:776]: I added the copy back

File: torch/_dynamo/variables/builder.py
Comment (jansel) [reply torch/_dynamo/variables/builder.py:473]: Perhaps add a helper method on `VariableBuilder` that sets the source and track side effects.

File: torch/_dynamo/side_effects.py
Comment (jansel) [inline torch/_dynamo/side_effects.py:521]:
```suggestion
                    cg.append_output(create_instruction("POP_TOP"))
```

File: torch/_dynamo/variables/builder.py
Comment (jansel) [inline torch/_dynamo/variables/builder.py:782]: Are all the source changes still needed with set_source_and_track_mutable?  (Throughout the file)

File: torch/_dynamo/variables/builder.py
Comment (jansel) [inline torch/_dynamo/variables/builder.py:245]: Is self.options() still used?  If not lets remove?

File: torch/_dynamo/variables/base.py
Comment (jansel) [inline torch/_dynamo/variables/base.py:349]: Is VT.clone() still used?  If not can we remove it?

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/variables/base.py
Comment (mlazos) [reply torch/_dynamo/variables/base.py:349]: I'll do this in a separate PR, I'm also going to remove the rest of the references to options (there are still some scattered about, but I want to isolate these changes for testing)

**Pull request closed without merging**
